### PR TITLE
Tracks tail update fix beyond _max_length

### PR DESCRIPTION
**Pull request by kushaangupta**
# Description

- Improve the intuitiveness while trying to set `tail_length` attribute of Tracks layer to beyond the hard set `_max_length`
- As a follow-up from suggestions by @alisterburt either we can auto-update `_max_length` upon change in `tail_length` or warn the user if they exceed `_max_length`.

- I chose the former as for the latter option I had to do the reassignment in this order to make it update in the viewer: `_max_length` -> `tail_length` -> again `tail_length`. It doesn't just seem to apply by just doing one reassignment of `tail_length`.
    - This change however seems to render `_max_length` of no use (need more discussion).
    - Along the same lines, `head_length` (which also uses `_max_length`) & `tail_width` (unable to update beyond 20) updates need to be considered

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/layers/tracks/tracks.py
Comment (Czaki) [inline napari/layers/tracks/tracks.py:464]: here you update class variable. You should not do this as it will impact all class instances. 

To fix this you need to at least move `_max_length` definition to constructor.

Comment (kushaangupta) [reply napari/layers/tracks/tracks.py:464]: Thank for catching that! I've updated it accordingly. In my opinion & @brisvag's maybe we can do without using `_max_length` & `_max_width` altogether

Comment (tlambert03) [reply napari/layers/tracks/tracks.py:464]: > here you update class variable.

fwiw, I'm not sure that's true in this case, because he did `self.___`:

```python
In [1]: class T:
   ...:     x = 1
   ...:

In [2]: t1 = T()

In [3]: t2 = T()

In [4]: t1.x
Out[4]: 1

In [5]: t2.x
Out[5]: 1

In [6]: T.x
Out[6]: 1

In [7]: t1.x = 2

In [8]: t1.x
Out[8]: 2

In [9]: t2.x
Out[9]: 1

In [10]: T.x
Out[10]: 1

In [11]: T.x = 4

In [12]: t2.x
Out[12]: 4

In [13]: t1.x
Out[13]: 2

In [14]: del t1.x

In [15]: t1.x
Out[15]: 4
```

(it's tricky :) but you don't _need_ to put it in the constructor here)

Comment (Czaki) [reply napari/layers/tracks/tracks.py:464]: I see a single `_` but maybe this is some part of python tricks and may be problematic to many contributors. I'm pretty sure that I see a scenario where it modifies the class variable.  

I still prefer to move this to the constructor.

**Comment by Czaki** [comment]
@kushaangupta I really like this change but it will be nice to add a test to this PR to prevent breaking this in future commits.

File: napari/layers/tracks/tracks.py
Comment (tlambert03) [reply napari/layers/tracks/tracks.py:464]: no it's not about the underscore, the point is that attribute access happens _first_ on the object, and then falls back to the class.  So, if you don't use `self.something = 'seomthing'` then it will get the class attribute, but once you set `self.something`... then you're setting it on the instance, not on the class.  
In any case, I agree, it's tricky, and it's just fine to put it in the constructor.  But didn't want anyone to get the wrong idea about how attribute access works :)

**Review by tlambert03 — APPROVED** [review]

**Review by alisterburt — APPROVED** [review]
Great stuff @kushaangupta ! Thanks for taking the time to make a PR, you're helping make napari better for everyone 🙂

Edit: I will add a test and merge this in a few hours unless anything comes up before then

**Comment by kushaangupta** [comment]
Thanks for the warm welcome and help to everyone! I did the same thing for `head_length` as for `tail_length`, maybe we can discuss what to do with the `tail_width` property also. Moreover, I wonder why there's no `head_width`.
I didn't had any idea where to put test cases for tracks layer so I couldn't update it.

**Comment by andy-sweet** [comment]
This has two approvals and the changes look good enough to me, so I'm tempted to merge this soon.

@kushaangupta : thanks for the contribution here!

My only remaining thought is that `_max_length` doesn't do much now other than set the upper limit of the GUI control for the head and tail lengths (in `qt_tracks_controls.py`). If that's all it does, I think it would be better defined/indicated as a constant in the layer (that is not modified by the setters) or maybe better just a literal `int` in `QTracksControls.__init__` given that it's clear from the context what that number means.

Alternatively, if we intend `_max_length` to be mutable as in this PR, maybe we should sync the maximum value of the slider in `QTrackControls` to sync with that? We have similar existing problems (e.g. `Points.size`), so that's probably better handled outside of this PR.

> maybe we can discuss what to do with the `tail_width` property also.

I'm not the most familiar with the tracks layer, but it looks like `tail_width` should probably be renamed `width` as it controls the width of the head and tail of the tracks (which seems like it should cover most usage). No need to update that name in this PR though.

**Comment by alisterburt** [comment]
I'm with @andy-sweet - going to merge this now, thanks again for the PR @kushaangupta !! 

@andy-sweet I think I would prefer to see the max length for the GUI be set in `QTracksControls` and some API for getting access to the controls should it be needed, what do you think? Agree it's out of scope for this PR

`head_length` was only added relatively recently, I agree `tail_width` should be updated, will open an issue now

**Pull request merged**
